packages/streaming-image-volume-loader/src/cornerstoneStreamingImageVolumeLoader.ts

```diff
@@ -36,140 +42,186 @@ function cornerstoneStreamingImageVolumeLoader(
+    const sizeInBytesPerComponent =
+      bytesPerVoxel * dimensions[0] * dimensions[1] * dimensions[2];
+
+    let numComponents = 1;
+    if (PhotometricInterpretation === 'RGB') {
+      numComponents = 3;
+    }
+
+    const sizeInBytes = sizeInBytesPerComponent * numComponents;
+
+    // check if there is enough space in unallocated + image Cache
+    const isCacheable = cache.isCacheable(sizeInBytes);
+    if (!isCacheable) {
+      throw new Error(Enums.Events.CACHE_SIZE_EXCEEDED);
+    }
+
+    cache.decacheIfNecessaryUntilBytesAvailable(sizeInBytes);
+
+    let scalarData;
+
+    switch (BitsAllocated) {
+      case 8:
+        if (signed) {
+          throw new Error(
+            '8 Bit signed images are not yet supported by this plugin.'
+          );
+        } else {
+          scalarData = createUint8SharedArray(
+            dimensions[0] * dimensions[1] * dimensions[2]
+          );
+        }
+
+        break;
+
+      case 16:
+        scalarData = createFloat32SharedArray(
           dimensions[0] * dimensions[1] * dimensions[2]
         );
-      }
 
-      break;
+        break;
 
-    case 16:
-      scalarData = createFloat32SharedArray(
-        dimensions[0] * dimensions[1] * dimensions[2]
-      );
+      case 24:
+        // hacky because we don't support alpha channel in dicom
+        scalarData = createUint8SharedArray(
+          dimensions[0] * dimensions[1] * dimensions[2] * numComponents
+        );
 
-      break;
+        break;
+    }
 
-    case 24:
-      // hacky because we don't support alpha channel in dicom
-      scalarData = createUint8SharedArray(
-        dimensions[0] * dimensions[1] * dimensions[2] * numComponents
-      );
+    let streamingImageVolume = new StreamingImageVolume(
+      // ImageVolume properties
+      {
+        volumeId,
+        metadata: volumeMetadata,
+        dimensions,
+        spacing,
+        origin,
+        direction,
+        scalarData,
+        sizeInBytes,
+      },
+      // Streaming properties
+      {
+        imageIds: sortedImageIds,
+        loadStatus: {
+          // todo: loading and loaded should be on ImageVolume
+          loaded: false,
+          loading: false,
+          cachedFrames: [],
+          callbacks: [],
+        },
+      }
+    );
 
-      break;
+    return streamingImageVolume;
   }
 
-  let streamingImageVolume = new StreamingImageVolume(
-    // ImageVolume properties
-    {
-      volumeId,
-      metadata: volumeMetadata,
-      dimensions,
-      spacing,
-      origin,
-      direction,
-      scalarData,
-      sizeInBytes,
-    },
-    // Streaming properties
-    {
-      imageIds: sortedImageIds,
-      loadStatus: {
-        // todo: loading and loaded should be on ImageVolume
-        loaded: false,
-        loading: false,
-        cachedFrames: [],
-        callbacks: [],
-      },
-    }
-  );
-
   return {
-    promise: Promise.resolve(streamingImageVolume),
+    promise: getStreamingImageVolume(),
```

Comment:
can't we say before return something like `const streamingImageVolume = await getStreaming .... ` and then use it below directly like before?

Reply:
Yes -- how about `const streamingImageVolumePromise = getStreaming...` so that it's explicit? Will defer to your preference.